src/track/globaltrackcache.h

```diff
@@ -156,24 +189,29 @@ class GlobalTrackCache {
             const QFileInfo& fileInfo,
             const SecurityTokenPointer& pSecurityToken = SecurityTokenPointer());
 
-    void evictAll();
+    bool isEmpty() const;
+
+    // Reset all indices but keep the currently allocated tracks
+    // to prevent memory leaks.
+    void resetIndices();
 
 private:
     friend class GlobalTrackCacheLocker;
     friend class GlobalTrackCacheResolver;
 
     static GlobalTrackCache* volatile s_pInstance;
 
+    // Callback for the smart-pointer
     static void deleter(Track* pTrack);
 
     class Item final {
     public:
         Item()
             : plainPtr(nullptr) {
         }
-        Item(const TrackRef trackRef,
+        Item(TrackRef trackRef,
                 const TrackPointer& pTrack)
-            : ref(trackRef),
+            : ref(std::move(trackRef)),
               weakPtr(pTrack),
               plainPtr(pTrack.get()) {
```

Comment:
We should get rid of plainPtr, because it may point to an invalid memory once the owning last TrackPointer falls out of scope.

Reply:
Nope, not possible. The plain pointer is essential for reviving entries. And reviving entries is required to resolve race conditions.

Reply:
IMHO we must not revive a track from a plain pointer. This breakers the shared / weak pointer safety.
If we want to keep a track in cache it has to become a shared pointer. If not, it is a weak pointer.

Reply:
Once again, not possible. Unfortunately it is not that simple. Our custom deleter of shared_ptr is invoked without locking and we have to compensate for those race conditions that occur before entering the cache. I understand that this is mind-boggling and it took me a huge amount of time to recognize what is actually happening.

If we don't revive tracks we will lose modified metadata due to race conditions. New instances would be created before having the chance to save existing ones where the reference count just dropped to zero. It all starts with and depends on the race condition I outlined in the comments of  GlobalTrackCache::deleter().

Reply:
Ah, I got it. 
Unfortunately, the current implementation is hard to understand, and probably not save, for future changes. The Smart pointers are kind of "missused" compared to their intended use case. 

The critical situation now, is when the last shared pointer falls out of scope, but the changes are not committed to the file or the database, right? 
It like be great if we could trigger the write back at reference 1. 

I have to read the code carefully again.
Maybe we can use a second smart pointer for storing the changes, or we can inherit from the smart pointer and add a last reference hook or something.

Reply:
A reference count of 1 is not sufficient for triggering any pre-mortal actions and we also don't get any callback for this event.

The code is far too complex and the almost paranoid usage of debug assertions is the icing on the cake. We have to be very careful when refactoring it. I don't think that I will be able to get it much more readable until the release.